sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/FilterEstimation.scala

```diff
@@ -0,0 +1,618 @@
+        // We update ColumnStat structure after apply this equality predicate.
+        // Set distinctCount to 1.  Set nullCount to 0.
+        val newStats = attrRef.dataType match {
+          case _: NumericType =>
+            val newValue = Some(literal.value)
+            aColStat.copy(distinctCount = 1, min = newValue,
+              max = newValue, nullCount = 0)
+          case DateType =>
+            val dateValue = literal.dataType match {
+              case StringType =>
+                Some(Date.valueOf(literal.value.toString))
+              case _ => Some(DateTimeUtils.toJavaDate(literal.value.toString.toInt))
+            }
+            aColStat.copy(distinctCount = 1, min = dateValue,
+              max = dateValue, nullCount = 0)
+          case TimestampType =>
+            val tsValue = literal.dataType match {
+              case StringType =>
+                Some(Timestamp.valueOf(literal.value.toString))
+              case _ => Some(DateTimeUtils.toJavaTimestamp(literal.value.toString.toLong))
+            }
+            aColStat.copy(distinctCount = 1, min = tsValue,
+              max = tsValue, nullCount = 0)
+          case _ => aColStat.copy(distinctCount = 1, nullCount = 0)
+        }
+        mutableColStats += (attrRef.exprId -> newStats)
+      }
+
+      Some(1.0 / ndv.toDouble)
+    } else {
+      Some(0.0)
+    }
+
+  }
+
+  /**
+   * Returns a percentage of rows meeting "IN" operator expression.
+   * This method evaluates the equality predicate for all data types.
+   *
+   * @param attrRef an AttributeReference (or a column)
+   * @param hSet a set of literal values
+   * @param update a boolean flag to specify if we need to update ColumnStat of a given column
+   *               for subsequent conditions
+   * @return a doube value to show the percentage of rows meeting a given condition
+   *         It returns None if no statistics exists for a given column.
+   */
+
+  def evaluateInSet(
+      attrRef: AttributeReference,
+      hSet: Set[Any],
+      update: Boolean)
+    : Option[Double] = {
+    if (!mutableColStats.contains(attrRef.exprId)) {
+      logDebug("[CBO] No statistics for " + attrRef)
+      return None
+    }
+
+    val aColStat = mutableColStats(attrRef.exprId)
+    val ndv = aColStat.distinctCount
+    val aType = attrRef.dataType
+
+    // use [min, max] to filter the original hSet
+    val validQuerySet = aType match {
+      case _: NumericType | DateType | TimestampType =>
+        val statsRange =
+          Range(aColStat.min, aColStat.max, aType).asInstanceOf[NumericRange]
+        hSet.map(e => numericLiteralToBigDecimal(e, aType, isNumeric = true))
+          .filter(e => e >= statsRange.min && e <= statsRange.max)
+
+      // We assume the whole set since there is no min/max information for String/Binary type
+      case StringType | BinaryType => hSet
+    }
+    if (validQuerySet.isEmpty) {
+      return Some(0.0)
+    }
+
+    val newNdv = validQuerySet.size

```

Comment:
newNdv can NOT be larger than the old ndv. e.g. col in (1, 2, 3, 4, 5), if col has only two values 1, 6, original ndv is 2, but validQuerySet.size is 5